src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/ConvertToRecordHelpers.cs

```diff
@@ -0,0 +1,914 @@
+                    {
+                        DeclaredSymbol: ISymbol otherVar,
+                        MatchedType: INamedTypeSymbol matchedType,
+                    } &&
+                    matchedType.Equals(currentObject.GetSymbolType()) &&
+                    // found the correct binding, add any members we equate in the rest of the binary condition
+                    // if we were in a binary condition at all, and signal failure if any condition is bad
+                    additionalConditions.All(otherCondition => TryAddEqualizedFieldsForCondition(
+                        otherCondition, successRequirement, currentObject, otherVar, builder)))
+                {
+                    boundVariable = otherVar;
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Match a list of statements and add members that are compared
+        /// </summary>
+        /// <param name="statementsToCheck">operations which should compare members</param>
+        /// <param name="otherC">non-this comparison of the type we're equating</param>
+        /// <param name="type">the this symbol</param>
+        /// <param name="builder">builder for property/field comparisons we encounter</param>
+        /// <returns>whether every statment was one of the expected forms</returns>
+        private static bool TryAddEqualizedFieldsForStatements(
+            IEnumerable<IOperation> statementsToCheck,
+            ISymbol otherC,
+            INamedTypeSymbol type,
+            ArrayBuilder<IFieldSymbol> builder)
+            => statementsToCheck.FirstOrDefault() switch
+            {
+                IReturnOperation
+                {
+                    ReturnedValue: ILiteralOperation
+                    {
+                        ConstantValue.HasValue: true,
+                        ConstantValue.Value: true,
+                    }
+                }
+                    // we are done with the comparison, the final statment does no checks
+                    => true,
+                IReturnOperation { ReturnedValue: IOperation value } => TryAddEqualizedFieldsForCondition(
+                    value, successRequirement: true, currentObject: type, otherObject: otherC, builder: builder),
+                IConditionalOperation
+                {
+                    Condition: IOperation condition,
+                    WhenTrue: IOperation whenTrue,
+                    WhenFalse: var whenFalse,
+                }
+                    // 1. Check structure of if statment, get success requirement
+                    // and any potential statments in the non failure block
+                    // 2. Check condition for compared members
+                    // 3. Check remaining members in non failure block
+                    => TryGetSuccessCondition(whenTrue, whenFalse, statementsToCheck.Skip(1).AsImmutable(),
```

Comment:
So if we have a bunch of statements is this .Skip(1).AsImmutable going to be allocating O(n^2) ish?

Reply:
If I make all the parameters IEnumerable instead of Immutable arrays will that mean they don't need to get re-allocated? Or will it build up a ridiculous amount of queries? I think a linked list would be best here because most of my actions are adding/removing from the front or back, but I'm not sure its the most usable/interopable with the IOperation properties